File: Finished/symmetricTree.js
// https://leetcode.com/problems/symmetric-tree/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function (root) {
  if (root === null) {
    return true;
  }

  root.right = Mirror(root.right);

  return isSame(root.left, root.right);
};

function Mirror(root) {
  if (root === null) {
    return root;
  }

  const left = Mirror(root.left);
  const right = Mirror(root.right);
  root.left = right;
  root.right = left;
  return root;
}

function isSame(left, right) {
  if (left === null && right === null) {
    return true;
  }

  if (left === null || right === null) {
    return false;
  }

  if (left.val !== right.val) {
    return false;
  }

  return isSame(left.left, right.left) && isSame(left.right, right.right);
}

File: src/python/finished/symmetric_tree.py
from typing import Optional


# https://leetcode.com/problems/symmetric-tree/


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        if root is None:
            return True

        def mirror_compare(left, right) -> bool:
            if left == right:
                return True

            if left is None or right is None:
                return False

            if left.val != right.val:
                return False

            return mirror_compare(left.left, right.right) and mirror_compare(
                left.right, right.left
            )

        return mirror_compare(root.left, root.right)
